pkg/controller/cheinstallation/cheinstallation_controller.go

```diff
@@ -72,10 +75,18 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
-	if err := c.Watch(&source.Kind{Type: &orgv1.CheCluster{}}, enqueueRequestForOwner); err != nil {
-		return err
+	r.watchCheCluster = func() error {
+		return c.Watch(&source.Kind{Type: &orgv1.CheCluster{}}, enqueueRequestForOwner)
 	}
 
+	err = r.watchCheCluster()
+	if err != nil {
+		if _, ok := err.(*meta.NoKindMatchError); !ok { // ignore NoKindMatchError
```

Comment:
what does it mean that we skip the `NoKindMatchError`?  shouldn't we set the watch function to the param in the service?

Reply:
ah, wait, you have set it already on line 78... I would change the flow - firstly try to call the `watch`, then if it return the error `NoKindMatchError` assign it to the parameter `watchCheCluster`

Reply:
It's about re-use the add_watch logic rather duplicate at both places.  So all the logic stay with `watchCheCluster()` and we are calling it from two different places, one from ctrl_init() and other from reconcile() .. but same logic .. no duplication.

Reply:
I meant this flow:
```go
watchCheCluster := func() error {
	return c.Watch(&source.Kind{Type: &orgv1.CheCluster{}}, enqueueRequestForOwner)
}

err = watchCheCluster()
if err != nil {
	if _, ok := err.(*meta.NoKindMatchError); !ok { // ignore NoKindMatchError
		return err
	}
	r.watchCheCluster = watchCheCluster
}
```
* it has no duplication
* it's more intuitive and easy to read - "assign it only when the watch call returns `NoKindMatchError`
* the code looks simpler

Reply:
Done https://github.com/nurali-techie/toolchain-operator/pull/1/commits/eb8482409a961202a19c9ed15500bbb48e68dd03